Model comparison v2

tolga-b opened the pull request:
Second version of model comparison for what-if-tool plugin.

**Motivation for features / changes**
PR #1589 added the ability to load two models and compare their inference scores in Facets Dive panel for binary classification and Datapoint Editor tab for classification.
This PR extends the model comparison mode to Global/Partial PDP, Fairness and Performance panels. It also adds comparing regression and multi-class models in all supported panels.
All model prediction related features in facets dive (Inference Value, Inference Error etc.) now also have multiple model support.

The only mode that does not support multiple models is closest counterfactual button which still computes its results on model 1.

**Technical description of changes**
Some variables that used to handle single models became arrays indexed by model number.
Some string keys for facets features now have model number appended to them.

Add the ability to compare two models them for Global PDP, PDP plots:
Now the Global PDP and PDP plots show one line per model per label in the same chart.
(Previously they only showed one line per label.)

Performance and Fairness panel:
Show one ROC curve per model within the same chart. (Previously showed ROC curve for single model.)
Show one threshold and confusion matrix per model. (Previously showed threshold and confusion matrix for single model.)

Add the ability to load two models for regression:
Facets dive features and MSEs in Performance and Fairness panel work for two regression models.

Minor:
- Do not show model number in facets "Inference Score" when only one model is present.
- Categorical PD plots now show the original value as text under the bar (feature name - original value), instead of showing orange bar for original feature and blue for others
- Remove forceUpdateFeatureValueThresholds as it is unnecessary when bound to arrays with "{{}}" in dom-repeat
- Fixes a bug with multi-class button in inference settings panel where clicking on it after the tool is used on a binary model without refreshing the page could throw an exception.

**Screenshots of UI changes**
all features now support two models
![all features now support two models](https://user-images.githubusercontent.com/6536229/50362147-a2d7ca00-0534-11e9-9e48-1d03d65c7fc0.png)

features old
![features old](https://user-images.githubusercontent.com/6536229/50362160-ab300500-0534-11e9-881a-e95479b4d91e.png)

new performance and fairness tab
![new performance and fairness tab](https://user-images.githubusercontent.com/6536229/50362304-314c4b80-0535-11e9-8425-1a37702dccc7.png)

old performance and fairness tab
![old performance and fairness tab](https://user-images.githubusercontent.com/6536229/50362309-33aea580-0535-11e9-8df3-44b6d27047ba.png)

pdp shows multiple models
![pdp shows multiple models](https://user-images.githubusercontent.com/6536229/50362382-89834d80-0535-11e9-88cb-cc9b9191d44b.png)

same for multi-class case
![same for multi-class case](https://user-images.githubusercontent.com/6536229/50362447-ca7b6200-0535-11e9-8321-020510fa1334.png)

categorical pdp shows each model as separate bar and labels original value in text
![categorical pdp shows each model as separate bar and labels original value in text](https://user-images.githubusercontent.com/6536229/50362398-9738d300-0535-11e9-8e4b-f1eb27b82cf9.png)

all regression features are available in facets
![regression features](https://user-images.githubusercontent.com/6536229/50362642-77ee7580-0536-11e9-9081-0479daf80728.png)

regression performance shows both models
![regression performance](https://user-images.githubusercontent.com/6536229/50362645-7b81fc80-0536-11e9-9458-3ed68126a48d.png)

**Detailed steps to verify changes work correctly (as executed by you)**

Tested all offline demos, multi/single model for binary and multi-class models (using the settings panel multi-class option with binary models). 
Tested multiple model regression with age demo by duplicating the first model.
Testing involves visually verifying the functionality of all panels by clicking buttons and verifying the outputs and checking for exceptions in the browser console.

**Alternate designs / implementations considered**

N/A

jameswex commented on tensorboard/plugins/interactive_inference/utils/inference_utils.py:
```diff
@@ -509,6 +514,8 @@ def make_json_formatted_for_single_chart(mutant_features,
       of mutant_features.
     index_to_mutate: The index of the feature being mutated for this chart.
+    model_id: Index of the model corresponding to inference_results_proto to
+      distinguish between multiple models for model comparison
```
this isn't used anywhere in the method?

tolga-b replied:
you are right, removed it.

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -1292,14 +1291,16 @@ <h2>Create a distance feature</h2>
-                            <div class="slider-label">[[getPrintableValue_(item)]]</div>
+                            <div class="slider-label">[[getPrintableValue_(featureValueThreshold)]]</div>
                             <div class="flex">
+                              <template is="dom-repeat" items="{{featureValueThreshold.threshold}}">
```
indent the paper-slider

tolga-b replied:
done

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -1340,9 +1345,12 @@ <h2>Create a distance feature</h2>
-                                            immediate-value="{{overallThreshold}}" value="[[overallThreshold]]">
-                              </paper-slider>
+                              <template is="dom-repeat" items="{{overallThresholds}}" as="overallThreshold">
+                                <paper-slider class="slider" editable=true min="0" max="1" step="0.01"
```
add label for sliders when more than one model?

tolga-b replied:
added.

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -1373,28 +1381,36 @@ <h2>Create a distance feature</h2>
-                                                  class="conf-matrix"></tf-confusion-matrix>
-                              </tf-confusion-matrix>
+                              <div>
+                                <template is="dom-repeat" items="[[inferenceStats_]]" index-as="modelInd">
```
label conf matrices when more than one model

tolga-b replied:
added.

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -3662,59 +3815,121 @@ <h2>Create a distance feature</h2>
+        // series
+        self = this;
+        const mapped = {};
+        _.forEach(data, function(modelValues, modelInd){
```
i know the code was like this before, but if you remove function() {} and replace with () => {} here and below, then you don't need the self=this, can just use this.  also if you were to use "self", should be "const self"

tolga-b replied:
done.

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -2184,13 +2211,32 @@ <h2>Create a distance feature</h2>
+
+      getNumberOfModels_: function(){
+        return this.modelName.split(',').length;
+      },
```
its strange that each time we want this we have to do this string split. i think numModels should be a data member that is set/updated at any appropriate time during inference or after model name setting, and then is used without needing recalcuating.

tolga-b replied:
You are right, moved it to variable numModels which gets updated when user clicks Accept in settings pane.

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -2281,190 +2333,225 @@ <h2>Create a distance feature</h2>
+            plotStats.push(inferenceStats.faceted[key]);
+            plotThresholds.push(modelThresholds[modelInd].threshold)
+          }
+        this.plotPr(
```
indenting issue here maybe?

tolga-b replied:
done.

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -3012,29 +3152,35 @@ <h2>Create a distance feature</h2>
-            inferenceMap[this.inferences.indices[i]][0] =
+            for (let modelNum = 0; modelNum < this.inferences.results.length;
+                 modelNum++){
+            const result = this.inferences.results[modelNum].regressionResult;
```
indenting issue

tolga-b replied:
done.

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -3085,19 +3234,26 @@ <h2>Create a distance feature</h2>
+                modelInd < self.getNumberOfModels_();
+                modelInd++) {
+              item[self.strWithModelIndex_(inferenceLabelStr, modelInd)] =
+              self.labelVocab[
```
whitespace

tolga-b replied:
done.

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -3109,48 +3265,36 @@ <h2>Create a distance feature</h2>
 
         // Binary classifier case.
-        let threshold = this.overallThreshold;
+        let threshold = this.overallThresholds;
```
probably better if this is called "thresholds" as it is a list

tolga-b replied:
updated.

tolga-b commented:
Thank you James! Should be all set with the comments.

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -2295,194 +2360,231 @@ <h2>Create a distance feature</h2>
+            this.examplesAndInferences[0].inferences == null ||
+            this.examplesAndInferences[0].inferences[0].length
+              != this.numModels ||
+            (this.inferenceStats_.length != this.numModels &&
```
maybe add comments about this check, as its gotten complex with these new cases

tolga-b replied:
added description.

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -3025,34 +3187,46 @@ <h2>Create a distance feature</h2>
+        // console.log('overallthresholds', this.overallThresholds)
+        // console.log('inferencestats', this.inferenceStats_)
+        // console.log('inferences', this.inferences.results)
+        // console.log('examplesandinferences', this.examplesAndInferences)
```
remove commented-out debug  logs

tolga-b replied:
oops, done!

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -3680,59 +3863,120 @@ <h2>Create a distance feature</h2>
+          },
+        ];
+        // If there is more than 1 model, show model number tooltip
+        if (data.length >= 2){
```
nit: space before {

tolga-b replied:
done.

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -3680,59 +3863,120 @@ <h2>Create a distance feature</h2>
+      },
+
+      formatChartKey: function(labelKey, modelInd, numberOfModels) {
+          if (numberOfModels == 1){
```
space before {

tolga-b replied:
done.

tolga-b commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -467,28 +467,30 @@
+        bottom: 110px;
         transform: rotate(270deg);
+        font-size: 12px;
+        color: #3c4043;
```
Mahima added it yesterday.

tolga-b commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -222,8 +221,9 @@
       .pr-line-chart {
-        height: 150px;
-        width: 250px;
+        margin: 6px 6px 6px 6px;
```
thanks, done.

tolga-b commented on tensorboard/plugins/interactive_inference/utils/inference_utils.py:
```diff
@@ -463,11 +463,14 @@ def chart_for_index(index_to_mutate):
-        mutant_examples, serving_bundle)
-    return make_json_formatted_for_single_chart(mutant_features,
+    charts = []
+    for model_id, serving_bundle in enumerate(serving_bundles):
```
good catch, removed!

tolga-b commented:
Did another pass.
Tested all offline demos, multi/single model for binary and multi-class models (using the settings panel multi-class option with binary models). Tested multiple model regression with age demo by duplicating the first model. 

Added binary check for ROC curve plotting in updateInferenceStats. Before, it only checked for the existence of chart, but if someone switches from binary to multi-class using inference settings window without refreshing the page, the chart elements still exist in the dom which causes an error in plotPR (thresholds don't exist for non-binary inferenceStats.faceted[key] is an object not an array so map in plotPR throws an error).

Added "restamp" to binary/multi-class/regression dom-ifs to delete unused chunks when model type changes. There is currently an issue with multi-class selection in inference settings panel where one may trigger a number of thresholds x number of thresholds (100 x 100) confusion matrix when switching from binary to multi-class model without refreshing the page. Without "restamp" this cripples the GUI after that event. With "restamp", the confusion matrix gets cleared after creation so the issue is temporary. Deferring fixing this to a future PR as it is independent from model comparison.
